### PR TITLE
Add support for X.509 in demo server

### PR DESCRIPTION
**Pull request by scop**
Tested successfully with `output/bin/demo` and the generated certs from `output/certs` with current https://github.com/AVSystem/Anjay master. See also https://github.com/AVSystem/Anjay/pull/7

### CODE REVIEW

File: leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/LeshanServerDemo.java [hunk 1]
```diff
@@ -190,35 +221,89 @@ public static void createAndStartServer(int webPort, String localAddress, int lo
...
+                KeyStore keyStore = KeyStore.getInstance(keyStoreType);
+                try (FileInputStream fis = new FileInputStream(keyStorePath)) {
+                    keyStore.load(fis, keyStorePass == null ? null : keyStorePass.toCharArray());
+                    ArrayList<Certificate> trustedCertificates = new ArrayList<>();
+                    for (Enumeration<String> aliases = keyStore.aliases(); aliases.hasMoreElements();) {
+                        String alias = aliases.nextElement();
+                        if (keyStore.isCertificateEntry(alias)) {
+                            trustedCertificates.add(keyStore.getCertificate(alias));
+                        } else if (keyStore.isKeyEntry(alias) && alias.equals(keyStoreAlias)) {
+                            ArrayList<X509Certificate> x509CertificateChain = new ArrayList<>();
```
Comment (jvermillard) [inline leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/LeshanServerDemo.java]: List<...>

File: leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/LeshanServerDemo.java [hunk 2]
```diff
@@ -190,35 +221,89 @@ public static void createAndStartServer(int webPort, String localAddress, int lo
...
-            LOG.error("Unable to initialize RPK.", e);
-            System.exit(-1);
+
+        // Set up X.509 mode
+        if (keyStorePath != null) {
+            try {
+                KeyStore keyStore = KeyStore.getInstance(keyStoreType);
+                try (FileInputStream fis = new FileInputStream(keyStorePath)) {
+                    keyStore.load(fis, keyStorePass == null ? null : keyStorePass.toCharArray());
+                    ArrayList<Certificate> trustedCertificates = new ArrayList<>();
```
Comment (jvermillard) [inline leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/LeshanServerDemo.java]: List<...>

File: leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/LeshanServerDemo.java [hunk 3]
```diff
@@ -190,35 +221,89 @@ public static void createAndStartServer(int webPort, String localAddress, int lo
...
+                            trustedCertificates.add(keyStore.getCertificate(alias));
+                        } else if (keyStore.isKeyEntry(alias) && alias.equals(keyStoreAlias)) {
+                            ArrayList<X509Certificate> x509CertificateChain = new ArrayList<>();
+                            Certificate[] certificateChain = keyStore.getCertificateChain(alias);
+                            if (certificateChain != null) {
+                                for (Certificate certificate : certificateChain) {
+                                    if (certificate instanceof X509Certificate) {
+                                        x509CertificateChain.add((X509Certificate) certificate);
+                                    } else {
+                                        LOG.warn("Ignoring non-X.509 certificate in alias chain: {}", certificate);
```
Comment (jvermillard) [inline leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/LeshanServerDemo.java]: this is probably an error if you have a non-X509 cert in the cert-chain, so an exception or any fatal error is better

File: leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/LeshanServerDemo.java [hunk 4]
```diff
@@ -190,35 +221,89 @@ public static void createAndStartServer(int webPort, String localAddress, int lo
...
+                                    publicKey = keyStore.getCertificate(alias).getPublicKey();
+                                    builder.setPublicKey(publicKey);
+                                    builder.setCertificateChain(x509CertificateChain
+                                            .toArray(new X509Certificate[x509CertificateChain.size()]));
+                                } else {
+                                    LOG.error("Keystore alias must have a PrivateKey entry, was {}",
+                                            key == null ? null : key.getClass().getName());
+                                    System.exit(-1);
+                                }
+                            } else {
```
Comment (jvermillard) [inline leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/LeshanServerDemo.java]: move the error case before and put the nominal case outside the if

**Comment by jvermillard** [comment]
Thanks! Would be nice to have the command line you used for generating the key-store, so we can put it in the instructions.

File: leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/LeshanServerDemo.java [hunk 5]
```diff
@@ -190,35 +221,89 @@ public static void createAndStartServer(int webPort, String localAddress, int lo
...
+                                    }
+                                }
+                            }
+                            if (!x509CertificateChain.isEmpty()) {
+                                Key key = keyStore.getKey(alias,
+                                        keyStoreAliasPass == null ? new char[0] : keyStoreAliasPass.toCharArray());
+                                if (key instanceof PrivateKey) {
+                                    builder.setPrivateKey((PrivateKey) key);
+                                    publicKey = keyStore.getCertificate(alias).getPublicKey();
+                                    builder.setPublicKey(publicKey);
```
Comment (sbernard31) [inline leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/LeshanServerDemo.java]: You don't need set public key as explain in LeshanServer javadoc. We should add javadoc to builder too.

**Comment by scop** [comment]
I think this addresses all the comments. Did not have time to re-test yet though.

I don't have the exact commands I used to create the keystore. What I basically did was took serverKeyStore.jks from Leshan integration tests, and added root.crt generated during Anjay's build to it using Portecle (portecle.sf.net, shameless plug).

**Comment by scop** [comment]
(Tested briefly now, still works for me.)

File: leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/LeshanServerDemo.java [hunk 6]
```diff
@@ -194,35 +225,88 @@ public static void createAndStartServer(int webPort, String localAddress, int lo
...
+                                    keyStoreAliasPass == null ? new char[0] : keyStoreAliasPass.toCharArray());
+                            if (key instanceof PrivateKey) {
+                                builder.setPrivateKey((PrivateKey) key);
+                                publicKey = keyStore.getCertificate(alias).getPublicKey();
+                                builder.setCertificateChain(
+                                        x509CertificateChain.toArray(new X509Certificate[x509CertificateChain.size()]));
+                            } else {
+                                LOG.error("Keystore alias must have a PrivateKey entry, was {}",
+                                        key == null ? null : key.getClass().getName());
+                                System.exit(-1);
```
Comment (sbernard31) [inline leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/LeshanServerDemo.java]: move the error case before and put the nominal case outside the if

File: leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/LeshanServerDemo.java [hunk 7]
```diff
@@ -194,35 +225,88 @@ public static void createAndStartServer(int webPort, String localAddress, int lo
...
+                                LOG.error("Keystore alias must have a non-empty chain of X509Certificates.");
+                                System.exit(-1);
+                            }
+
+                            for (Certificate certificate : certificateChain) {
+                                if (certificate instanceof X509Certificate) {
+                                    x509CertificateChain.add((X509Certificate) certificate);
+                                } else {
+                                    LOG.error("Non-X.509 certificate in alias chain is not supported: {}", certificate);
+                                    System.exit(-1);
```
Comment (sbernard31) [inline leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/LeshanServerDemo.java]: move the error case before and put the nominal case outside the if

**Comment by scop** [comment]
Remaining comments addressed now.

**Pull request merged**